agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+  } else if (nr_php_zend_hash_key_is_numeric(hash_key)) {
+    ZVAL_LONG(key, (zend_long)nr_php_zend_hash_key_integer(hash_key));
+    key_str = nr_formatf("%ld", (long)Z_LVAL_P(key));
+  } else {
+    /*
+     * This is a warning because this really, really shouldn't ever happen.
+     */
+    nrl_warning(NRL_INSTRUMENT, "%s: unexpected key type", __func__);
+    key_str = nr_formatf("unsupported-key-type");
+  }
+  nr_php_zval_free(&key);
+  return key_str;
+}
+
+/*
+ * Purpose : Format value of $context array's  element as string
+ *
+ * Params  : zval value
+ * *
+ * Returns : A new string representing zval; caller must free
+ *
+ */
+static char* nr_monolog_fmt_context_value(zval* zv) {
+  char* val_str = NULL;
+  zval* zv_str = nr_php_zval_alloc();
+
+  ZVAL_DUP(zv_str, zv);
+  convert_to_string(zv_str);
+  val_str = nr_strdup(Z_STRVAL_P(zv_str));
+  nr_php_zval_free(&zv_str);
+
+  return val_str;
+}
+
+/*
+ * Purpose : Format an element of $context array as "key => value" string
+ *
+ * Params  : zval value, pointer to string buffer to store formatted output
+ * and hash key
+ *
+ * Side effect : string buffer is reallocated with each call.
+ *
+ * Returns : ZEND_HASH_APPLY_KEEP to keep iteration
+ *
+ */
+static int nr_monolog_fmt_context_item(zval* value,
+                                       char** strbuf,
+                                       zend_hash_key* hash_key TSRMLS_DC) {
+  NR_UNUSED_TSRMLS;
+  char* key = nr_monolog_fmt_context_key(hash_key);
+  char* val = nr_monolog_fmt_context_value(value);
+
+  char* kv_str = nr_formatf("%s => %s", key, val);
+  nr_free(key);
+  nr_free(val);
+
+  char* sep = nr_strlen(*strbuf) > 1 ? ", " : "";
+  *strbuf = nr_str_append(*strbuf, kv_str, sep);
+  nr_free(kv_str);
+
+  return ZEND_HASH_APPLY_KEEP;
+}
+
+/*
+ * Purpose : Iterate over $context array and format each element
+ *
+ * Params  : string buffer to store formatted output and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_fmt_context(char* strbuf,
+                                    HashTable* context TSRMLS_DC) {
+  strbuf = nr_str_append(strbuf, "[", "");
+
+  nr_php_zend_hash_zval_apply(context,
+                              (nr_php_zval_apply_t)nr_monolog_fmt_context_item,
+                              (void*)&strbuf TSRMLS_CC);
+
+  return nr_str_append(strbuf, "]", "");
+}
+
+/*
+ * Purpose : Convert $context argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_get_context(size_t argc, NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* context = nr_strdup("");
+  zval* context_arg = NULL;
+
+  if (3 > argc) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context not available", __func__);
+    goto return_context;
```

Comment:
I know `goto`s are present in other parts of our codebase, but is this something that we want to avoid adding more of? Or are we fine with it in simple contexts like this?

Reply:
I wanted to avoid multiple return points from this function and avoid re-allocating memory. But since this approach led to memory leak, I'll refactor it.

Reply:
@ZNeumann Upon further inspection it turns out this function does not have mem leak issue. See [this comment](https://github.com/newrelic/newrelic-php-agent/pull/489/files#r935845159). There I do not plan to refactor it at this time.